### ISSUE TITLE
EFI Stub: Booting uki from iPXE fails; Error registering initrd: Already started

### ISSUE THREAD

**Issue by xuxuck** [systemd/systemd#26723]
### systemd version the issue has been seen with

253

### Used distribution

Arch Linux

### Linux kernel version used

6.2.2-arch1-1

### CPU architectures issue was seen on

x86_64

### Component

systemd-stub

### Expected behaviour you didn't see

successful boot

### Unexpected behaviour you saw

Unified kernel image does not boot. After a second or two it returns `Error registering initrd: Already started`.

I did already discuss this at the iPXE repository. See: https://github.com/ipxe/ipxe/discussions/915#discussion-4942530

### Steps to reproduce the problem

Build a uki with the systemd efi stub and try to boot it with iPXE. 
In detail:

1. prepare a uki
2. make it available at http://some-local-ip/uki.efi
3. boot [ipxe.efi](https://boot.ipxe.org/ipxe.efi)
4. press `ctrl` + `b` to get a shell
5. type:
```
ifopen
chain http://some-local-ip/uki.efi
```

**Comment by medhefgo** [comment]
As already pointed out by https://github.com/ipxe/ipxe/commit/6a004be0cceab5d669eedb5a2e6ee2feab31d5bd: The kernel only looks for initrds on a fixed device path. But I disagree with the author of that commit: replacing the registered initrds sounds like a bad idea. We don't own it and therefore should not touch it.

Either get iPXE to recognize UKIs and not register the intird or teach the kernel to look for initrds on its EFI_Handle instead of a fixed device path.

@ardbiesheuvel What's the kernel's standpoint on this?

**Comment by mcb30** [comment]
@medhefgo There is no viable alternative that I can see to the approach taken in [ipxe/ipxe@6a004be](https://github.com/ipxe/ipxe/commit/6a004be0cceab5d669eedb5a2e6ee2feab31d5bd)

Device paths are required to be unique within a system.  Bootloaders are not.  It is therefore inevitable that a bootloader may find itself in a situation in which the fixed device path already exists.

Getting iPXE to special-case UKIs would be a retrograde step.  UKIs are supposed to look like standard UEFI binaries, and should not require special handling.

At the point that the systemd EFI stub wants to register the device path, it **does** conceptually own that handle, because it is at that point the definitive provider of the initrd content to the kernel that it is about to execute.  The correct path is therefore for the systemd EFI stub to add its `EFI_LOAD_FILE2_PROTOCOL` instance to the existing handle.  Installing protocols onto an existing handle is completely normal in the UEFI world.

(For improved robustness in the general case, the systemd EFI stub should uninstall any existing `EFI_LOAD_FILE2_PROTOCOL` instance found on the handle.  However, this wouldn't be necessary to get a successful boot in this particular case, since when using a UKI there would be no initrd content for iPXE to provide and so there will be no `EFI_LOAD_FILE2_PROTOCOL` installed on the handle.)

**Comment by ardbiesheuvel** [comment]
First of all, I would expect iPXE to only install the initrd device path onto a handle if there is initrd content being provided. I am not 100% sure what the 'magic' initrd amounts to in this respect, but if it means the initrd device path is always installed, and that LoadFile2() may fail when called as iPXE figures out that no initrd content is available after all, I don't think iPXE's approach makes sense.

Then, backing LoadFile2 with SimpleFileSystemProtocol is kind of redundant, although I understand it may be needed in some cases. However, going forward, the 'initrd=' command line interface for the EFI stub is always enabled for all EFI architectures in Linux, and accepts device paths in text representation, so it can support any file system that EFI can access. So perhaps we could make this dependent on EFI stub version < v1.1 at some point, and just pass the file reference to initrd= otherwise? (https://git.kernel.org/pub/scm/linux/kernel/git/efi/efi.git/tag/?h=efi-next-for-v6.2)

As far as UKI + iPXE is concerned: if iPXE only provides a LoadFile2 implementation if an actual initrd is being exposed, how should the EFI stub choose between the one provided by UKI and the one provided by iPXE? So in that sense, throwing an error here makes sense, because there is no obviously correct answer here.

If UKI decides that its initrd must always supersede the one provided by a previous boot stage, then uninstalling the existing one is reasonable, and unfortunately, I don't think there is a cleaner way to do this than locating the device path, and if it exists, locating the protocol interface and uninstalling it.

**Comment by medhefgo** [comment]
> Getting iPXE to special-case UKIs would be a retrograde step. UKIs are supposed to look like standard UEFI binaries, and should not require special handling.

But standard UEFI binaries aren't the kernel and shouldn't be started as if they were. You would similarly be unable to load an initrd if you start a UEFI shell with iPXE in the same way and then try to use the `inird` command as that implementation refuses to replace/unload one unless it was loaded by the shell itself.

**Comment by medhefgo** [comment]
You say bug, I say correct behavior. UKIs are not kernels, they're OS loaders. Do you really provide initrds to `bootmgfw.efi` too?

**Comment by mcb30** [comment]
> You say bug, I say correct behavior. UKIs are not kernels, they're OS loaders. Do you really provide initrds to `bootmgfw.efi` too?

Yes.  There is no special casing for Linux kernels: iPXE treats them as just UEFI executables.  `bootmgfw.efi` is totally unaffected if there happens to be a handle somewhere in the UEFI handle database that it neither knows nor cares about.  There will always be hundreds of such handles.

**Comment by mcb30** [comment]
We have worked around this systemd bug in iPXE commit https://github.com/ipxe/ipxe/commit/09e8a1540.  As long as the user has not downloaded any additional files alongside the UKI, then the faulty code path in the systemd EFI stub will not be reached.

**Comment by poettering** [comment]
I guess we can close this now. I am pretty sure I agree with @ardbiesheuvel though that ipxe should never have taken the handle if it has no initrd anyway.
